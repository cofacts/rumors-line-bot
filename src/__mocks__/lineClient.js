export default {
  verifyIDToken: jest.fn(),
  revokeNotifyToken: jest.fn(),
};
