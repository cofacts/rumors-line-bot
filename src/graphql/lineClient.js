import fetch from 'node-fetch';
import querystring from 'querystring';
import rollbar from 'src/lib/rollbar';

/**
 * Reference: https://developers.line.biz/en/reference/social-api/#response-5
 *
 * @param {string} idToken
 * @return {object} Response from API
 */
export async function verifyIDToken(idToken) {
  const URL = 'https://api.line.me/oauth2/v2.1/verify';
  const resp = await fetch(URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: querystring.stringify({
      id_token: idToken,
      client_id: process.env.LINE_LOGIN_CHANNEL_ID,
    }),
  });

  const result = await resp.json();

  if (resp.status !== 200) {
    rollbar.error(`[verifyIDToken] ${resp.status}: ${result.message}.`, {
      result,
    });
  }

  return result;
}

/**
 * Reference: https://notify-bot.line.me/doc/en/
 *
 * @param {string} lineNotifyToken
 */
export async function revokeNotifyToken(lineNotifyToken) {
  const URL = 'https://notify-api.line.me/api/revoke';
  const resp = await fetch(URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      Authorization: `Bearer ${lineNotifyToken}`,
    },
  });
  const result = await resp.json();

  if (resp.status !== 200) {
    rollbar.error(`[revokeNotifyToken] ${resp.status}: ${result.message}.`, {
      result,
    });
  }

  return result;
}
