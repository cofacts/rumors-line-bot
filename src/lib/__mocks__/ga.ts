const eventMock = jest.fn();
const sendMock = jest.fn();
const setMock = jest.fn();

const ga = Object.assign(
  jest.fn().mockImplementation(() => ({
    event: eventMock,
    send: sendMock,
    set: setMock,
    screenview: () => undefined,
  })),
  {
    clearAllMocks: () => {
      eventMock.mockClear();
      sendMock.mockClear();
      setMock.mockClear();
      ga.mockClear();
    },
    eventMock,
    sendMock,
    setMock,
  }
);

export default ga;
