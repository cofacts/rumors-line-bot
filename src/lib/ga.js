import ua from 'universal-analytics';
import { gaTitle } from './sharedUtils';

/**
 * Sends a screen view and returns the visitor
 *
 * @param {string} uuid - Visitor uuid, corresponding to `messageSource`, it will be userId, groupId, roomId
 * @param {string} state - The current state the LINE bot is in
 * @param {string} documentTitle - The article content (user input) currently in search
 * @param {string} messageSource - The event source { 'user' | 'group' | 'room' }
 * @returns {ua.Visitor} universal analytics visitor
 */
export default function ga(
  uuid,
  state = 'N/A',
  documentTitle = '',
  messageSource = 'user'
) {
  const visitor = ua(process.env.GA_ID, uuid, { strictCidFormat: false });

  // Ref: https://developers.google.com/analytics/devguides/collection/analyticsjs/screens
  // screenview needs both screen name and application name, or it will print error
  //
  visitor.screenview(state, 'rumors-line-bot');

  if (documentTitle) {
    visitor.set('dt', gaTitle(documentTitle));
  }

  visitor.set('cd1', messageSource);

  return visitor;
}
