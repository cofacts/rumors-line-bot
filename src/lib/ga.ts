import ua from 'universal-analytics';
import type { EventParams } from 'universal-analytics';
import { gaTitle } from './sharedUtils';
import { insertEventBatch } from './bq';
import type { EventBatch } from 'src/rumors-db/bq/events';
import rollbar from './rollbar';

/**
 * Sends a screen view and returns the visitor
 *
 * @param uuid - Visitor uuid, corresponding to `messageSource`, it will be userId, groupId, roomId
 * @param state - The current state the LINE bot is in
 * @param documentTitle - The article content (user input) currently in search
 * @param messageSource - The event source { 'user' | 'group' | 'room' }
 * @returns universal analytics visitor
 */
export default function ga(
  uuid: string,
  state = 'N/A',
  documentTitle = '',
  messageSource: 'user' | 'group' | 'room' = 'user'
) {
  const visitor = ua(process.env.GA_ID || '', uuid, { strictCidFormat: false });

  // Ref: https://developers.google.com/analytics/devguides/collection/analyticsjs/screens
  // screenview needs both screen name and application name, or it will print error
  //
  visitor.screenview(state, 'rumors-line-bot');

  if (documentTitle) {
    visitor.set('dt', gaTitle(documentTitle));
  }

  visitor.set('cd1', messageSource);

  let events: EventBatch['events'] = [];
  const extra: Record<string, unknown> = {};

  return {
    set(key: string, value: unknown) {
      extra[key] = value;
      return visitor.set(key, value);
    },

    event(evt: EventParams) {
      events.push({
        category: evt.ec ?? null,
        action: evt.ea ?? null,
        label: evt.el ?? null,
        value: evt.ev === undefined ? null : +evt.ev,
        time: new Date(),
      });
      return visitor.event(evt);
    },
    send() {
      insertEventBatch({
        text: documentTitle,
        messageSource,
        events,
        createdAt: new Date(),
        extra,
      }).catch((e) => {
        console.error('[insertAnalytics]', e);
        rollbar.error(`[insertAnalytics] ${e.message}`, e);
      });
      events = [];
      return visitor.send();
    },
  };
}
