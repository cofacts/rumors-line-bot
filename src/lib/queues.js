import Bull from 'bull';

export const groupEventQueue = new Bull('groupEventQueue', {
  redis: process.env.REDIS_URL || 'redis://127.0.0.1:6379',
  // limiter: { max: 600, duration: 10 * 1000 },
});
export const expiredGroupEventQueue = new Bull('expiredGroupEventQueue', {
  redis: process.env.REDIS_URL || 'redis://127.0.0.1:6379',
  // limiter: { max: 600, duration: 10 * 1000 },
});
