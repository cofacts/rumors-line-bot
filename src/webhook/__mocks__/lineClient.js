const lineClient = {
  post: jest.fn(),
  get: jest.fn(),
  getContent: jest.fn(),
};
export default lineClient;
