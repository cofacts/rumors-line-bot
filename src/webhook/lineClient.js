import fetch from 'node-fetch';
import rollbar from 'src/lib/rollbar';

export default { post, get, getContent };

async function post(endpoint = '', body = {}, options = {}) {
  const URL = `https://api.line.me/v2/bot${endpoint}`;
  const resp = await fetch(URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${process.env.LINE_CHANNEL_TOKEN}`,
      ...options.headers,
    },
    ...options,
    body: JSON.stringify(body),
  });

  const result = await resp.json();

  if (resp.status !== 200) {
    console.error(JSON.stringify(result, null, '  '));

    rollbar.error(
      `[LINE Client] ${resp.status}: ${result.message}.`,
      {
        // Request object for rollbar server SDK
        headers: options.headers,
        body: JSON.stringify(body),
        url: URL,
        method: 'POST',
      },
      { result }
    );
  }

  return result;
}

async function get(endpoint = '', options = {}) {
  const URL = `https://api.line.me/v2/bot${endpoint}`;
  const resp = await fetch(URL, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${process.env.LINE_CHANNEL_TOKEN}`,
      ...options.headers,
    },
    ...options,
  });

  const result = await resp.json();

  if (resp.status !== 200) {
    console.error(JSON.stringify(result, null, '  '));

    rollbar.error(
      `[LINE Client] ${resp.status}: ${result.message}.`,
      {
        // Request object for rollbar server SDK
        headers: options.headers,
        url: URL,
        method: 'GET',
      },
      { result }
    );
  }

  return result;
}

async function getContent(messageId, options = {}) {
  // this endpoint is for sending and receiving large amounts of data in the LINE platform for Messaging API.
  // https://developers.line.biz/en/reference/messaging-api/#get-content
  const URL = `https://api-data.line.me/v2/bot/message/${messageId}/content`;
  const resp = await fetch(URL, {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${process.env.LINE_CHANNEL_TOKEN}`,
      ...options.headers,
    },
    ...options,
  });

  if (resp.status !== 200) {
    const err = await resp.json();
    console.error(JSON.stringify(err, null, '  '));

    rollbar.error(
      `[LINE Client] ${resp.status}: ${err.message}.`,
      {
        // Request object for rollbar server SDK
        headers: options.headers,
        url: URL,
        method: 'GET',
      },
      { result: err }
    );
  }

  return resp;
}
