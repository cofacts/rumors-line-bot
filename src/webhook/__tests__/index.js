jest.mock('src/webhook/checkSignatureAndParse');
jest.mock('src/webhook/lineClient');
jest.mock('src/lib/redisClient');
jest.mock('src/webhook/handlers/tutorial');
jest.mock('src/lib/ga');

import Koa from 'koa';
import request from 'supertest';
import MockDate from 'mockdate';

import webhookRouter from '../';
import UserSettings from '../../database/models/userSettings';
import lineClient from 'src/webhook/lineClient';
import { groupEventQueue, expiredGroupEventQueue } from 'src/lib/queues';
import redis from 'src/lib/redisClient';
import {
  createGreetingMessage,
  createTutorialMessage,
} from 'src/webhook/handlers/tutorial';
import ga from 'src/lib/ga';

const sleep = async (ms) => new Promise((resolve) => setTimeout(resolve, ms));

describe('Webhook router', () => {
  beforeEach(() => {
    redis.set.mockClear();
    lineClient.post.mockClear();
    createGreetingMessage.mockClear();
    createTutorialMessage.mockClear();
    ga.clearAllMocks();
  });

  beforeAll(async () => {
    MockDate.set(612921600000);

    if (await UserSettings.collectionExists()) {
      await (await UserSettings.client).drop();
    }
  });

  beforeEach(async () => {
    process.env.RUMORS_LINE_BOT_URL = 'https://testlinebot.cofacts';
  });

  afterAll(async () => {
    MockDate.reset();
    delete process.env.RUMORS_LINE_BOT_URL;
    await groupEventQueue.close();
    await expiredGroupEventQueue.close();
  });

  it('singleUserHandler() should ignore sticker messages', async () => {
    const userId = 'U4af4980630';
    const app = new Koa();
    app.use(webhookRouter.routes(), webhookRouter.allowedMethods());

    const eventObject = {
      events: [
        {
          replyToken: 'nHuyWiB7yP5Zw52FIkcQobQuGDXCTA',
          type: 'messages',
          mode: 'active',
          timestamp: 1462629479859,
          source: {
            type: 'user',
            userId,
          },
          message: {
            id: '325708',
            type: 'sticker',
            packageId: '1',
            stickerId: '1',
            stickerResourceType: 'STATIC',
          },
        },
      ],
    };

    const server = app.listen();

    await request(server).post('/').send(eventObject).expect(200);

    /**
     * The HTTP response isn't guaranteed the event handling to be complete
     */
    await sleep(500);

    // snapshot reply messages
    expect(lineClient.post.mock.calls).toMatchSnapshot();
    // snapshot user context
    expect(redis.set.mock.calls).toMatchSnapshot();

    return new Promise((resolve, reject) => {
      server.close((error) => {
        if (error) return reject(error);
        resolve();
      });
    });
  });
});
