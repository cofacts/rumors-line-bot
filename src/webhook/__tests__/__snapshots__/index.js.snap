// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webhook router singleUserHandler() should ignore sticker messages 1`] = `Array []`;

exports[`Webhook router singleUserHandler() should ignore sticker messages 2`] = `Array []`;
